01_classify_stratified/nbconverted/plot_cancer_type_results.py

```diff
@@ -283,3 +285,136 @@ def label_points(x, y, gene, sig, ax):
 plt.title('Cancer type confusion matrix, methylation data', size=14, pad=14)
 plt.tight_layout()
 
+
+# How to read the confusion matrices:
+# 
+# * Remember, we're doing one-vs-rest classification of the cancer types. The cancer type on the y-axis is what the model is trained to predict (i.e. the positive label in the training set), and the cancer type on the x-axis is what we're evaluating on (i.e. the true label in the test set).
+# 
+# * A high off-diagonal value in the heatmap indicates that a model trained to distinguish the y-axis cancer type from all the others often "confuses" it for the x-axis cancer type (i.e. they have similar gene expression profiles). Diagonal values indicate the accuracy of the model on the target cancer type (these values should be relatively high).
+# 
+# In the resulting heatmaps, we can observe a few things:
+# 
+# * The models are accurate overall (high on-diagonal values and mostly low off-diagonal values), and the "inaccuracies" (high off-diagonal values) we see aren't surprising based on our knowledge of relationships between cancer types.
+# 
+# * Cancer type predictions using methylation data are noisier and not quite as accurate compared to gene expression data.
+
+# ## Hierarchical clustering analysis
+# 
+# We want to see if these confusion matrices lead to reasonable "cancer groups" (i.e. are cancer types that we would expect to be similar close by in the "similarity matrix" given by model confusion?). To do this, we'll just do a super simple agglomerative hierarchical clustering.
+
+# In[14]:
+
+
+# cluster and plot dendrogram
+from scipy.cluster.hierarchy import dendrogram
+from sklearn.cluster import AgglomerativeClustering
+
+# https://github.com/scikit-learn/scikit-learn/blob/70cf4a676caa2d2dad2e3f6e4478d64bcb0506f7/examples/cluster/plot_hierarchical_clustering_dendrogram.py
+def plot_dendrogram(model, **kwargs):
+    
+    # Distances between each pair of children
+    # Since we don't have this information, we can use a uniform one for plotting
+    distance = np.arange(model.children_.shape[0])
+
+    # The number of observations contained in each cluster level
+    no_of_observations = np.arange(2, model.children_.shape[0]+2)
+
+    linkage_matrix = np.column_stack([model.children_,
+                                      distance,
+                                      no_of_observations]).astype(float)
+
+    # Plot the corresponding dendrogram
+    dendrogram(linkage_matrix, **kwargs)
+    
+# fit agglomerative clustering model to confusion (similarity) matrix
+model = AgglomerativeClustering(distance_threshold=0,
+                                n_clusters=None,
+                                affinity='precomputed',
+                                linkage='complete')
+# takes a distance matrix (1 - similarity)
+model = model.fit(1 - preds_expression_df.values)
+
+plot_dendrogram(model, labels=preds_expression_df.index.values)
```

Comment:
Whats your takeaway on the grouping? Cancers that group together are those that have similar methylation or expression performance patterns, meaning that the model trained on UVM has a similar performance compared to the model trained on SKCM? If the models perform similarly, what does that mean in terms of how you can use this information?

Reply:
Hopefully my first comment on what the matrix means makes this a bit clearer. Cancers that group together are ones the model tends to mistake for one another (UVM and SKCM are both types of melanoma, so it makes sense that they have similar expression and methylation, and a model trained to differentiate one from all other cancer types will sometimes mistakenly identify the other).

So it's not just about the model performance in general, but about the specific instances where the model makes mistakes, which can tell us something (in a bit of a roundabout way) about which cancers are similar.

We're going to try some more direct methods of calculating "distance" between cancer types in the future, so maybe that analysis will be a bit more intuitive than this one. It's not typical to use confusion matrices in this way, so it's totally understandable that things are unclear to you.